File: backend/app/services/cases.py
"""cases 도메인 읽기 서비스 — GET /api/v1/cases(R2.3). docs/DB_SCHEMA.md §4.3.

get_case_out은 다른 도메인(예: briefings)이 그대로 재사용할 수 있도록 이름·시그니처를
유지한다 — 변경 시 호출부가 깨진다.
"""

from __future__ import annotations

import datetime as dt

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models.approval import Approval
from app.models.case import Case, NextAction
from app.models.document import WorkerDocument
from app.models.worker import Worker
from app.schemas.case import (
    ApprovalChecklistItemOut,
    CaseDetailOut,
    CaseOut,
    CheckedItemOut,
    NextActionOut,
    PendingApprovalOut,
    WorkerDocumentOut,
    WorkerRefOut,
)
from app.services.approvals import usable_citation_count

# 목록 정렬 기준: 심각도(높을수록 먼저) → 마감일(빠를수록 먼저) → 생성순.
_SEVERITY_ORDER = {"CRITICAL": 0, "HIGH": 1, "MEDIUM": 2, "LOW": 3}


def _next_action_out(next_action: NextAction | None) -> NextActionOut | None:
    if next_action is None:
        return None
    return NextActionOut(
        action_id=next_action.id,
        label=next_action.label,
        state=next_action.state,
        requires_approval=next_action.requires_approval,
        kind=next_action.kind,
    )


def get_case_out(db: Session, company_id: str, case: Case) -> CaseOut:
    """이미 로드된 Case ORM 객체 1건을 worker + primary/secondary next_action과 조립한다.

    worker는 case.worker_id(있는 경우)로, next_action은 slot='primary'/'secondary'로
    각 1건 조회한다 — 둘 다 company_id로 스코프한다.
    """
    worker = None
    if case.worker_id is not None:
        worker = db.execute(
            select(Worker).where(Worker.company_id == company_id, Worker.id == case.worker_id)
        ).scalar_one_or_none()

    primary = db.execute(
        select(NextAction).where(
            NextAction.company_id == company_id,
            NextAction.case_id == case.id,
            NextAction.slot == "primary",
        )
    ).scalar_one_or_none()
    secondary = db.execute(
        select(NextAction).where(
            NextAction.company_id == company_id,
            NextAction.case_id == case.id,
            NextAction.slot == "secondary",
        )
    ).scalar_one_or_none()

    return CaseOut(
        id=case.id,
        case_code=case.case_code,
        title=case.title,
        severity=case.severity,
        state=case.state,
        agent_stage=case.agent_stage,
        due_date=case.due_date,
        approval_required=case.approval_required,
        prepared_by=case.prepared_by,
        prepared_run_id=case.prepared_run_id,
        worker=WorkerRefOut.model_validate(worker) if worker is not None else None,
        primary_action=_next_action_out(primary),
        secondary_action=_next_action_out(secondary),
    )


def _pending_approval_out(approval: Approval | None) -> PendingApprovalOut | None:
    if approval is None:
        return None
    checklist = None
    if approval.checklist is not None and isinstance(approval.checklist, list):
        checklist = [ApprovalChecklistItemOut(**item) for item in approval.checklist]
    return PendingApprovalOut(
        id=approval.id, action_id=approval.action_id, checklist=checklist, requested_at=approval.requested_at
    )


def _checked_items_out(checked_items: object) -> list[CheckedItemOut]:
    if not isinstance(checked_items, list):
        return []
    return [CheckedItemOut(**item) for item in checked_items]


def _worker_documents_out(db: Session, company_id: str, worker_id: str | None) -> list[WorkerDocumentOut]:
    """worker_id가 없는 케이스(예: 커맨드 런 기원)는 워커 엔티티가 없으므로 빈 배열 —
    SD-6, mock CaseSheet.docs와 동일 개념."""
    if worker_id is None:
        return []
    documents = (
        db.execute(
            select(WorkerDocument)
            .where(WorkerDocument.company_id == company_id, WorkerDocument.worker_id == worker_id)
            .order_by(WorkerDocument.created_at, WorkerDocument.id)
        )
        .scalars()
        .all()
    )
    return [
        WorkerDocumentOut(doc_type=d.doc_type, status=d.status, due_date=d.due_date, expires_at=d.expires_at)
        for d in documents
    ]


def get_case_detail_out(db: Session, company_id: str, case: Case) -> CaseDetailOut:
    """GET /api/v1/cases/{case_id} 전용 조립 — get_case_out(목록용) 위에 승인 화면·케이스
    시트 화면(SD-6)이 필요로 하는 필드를 얹는다(R2.4 Blocker A+B 해소: 프론트가 mock
    CASE_SHEETS 대신 이 응답으로 체크리스트·근거수·가드노트·pending approval_id를 얻는다.
    SD-6은 여기에 checked_items·next_wake(컬럼 그대로 노출)·worker_documents를 더한다)."""
    base = get_case_out(db, company_id, case)
    pending = db.execute(
        select(Approval).where(
            Approval.company_id == company_id, Approval.case_id == case.id, Approval.status == "pending"
        )
    ).scalar_one_or_none()
    return CaseDetailOut(
        **base.model_dump(),
        usable_citation_count=usable_citation_count(db, company_id, case.id),
        guard_note=case.guard_note,
        pending_approval=_pending_approval_out(pending),
        checked_items=_checked_items_out(case.checked_items),
        next_wake=case.next_wake_condition,
        documents=_worker_documents_out(db, company_id, case.worker_id),
    )


def list_cases_out(db: Session, company_id: str) -> list[CaseOut]:
    """company_id로 스코프된 전체 케이스를 심각도·마감일 순으로 조립해 반환한다."""
    cases = db.execute(select(Case).where(Case.company_id == company_id)).scalars().all()
    cases_sorted = sorted(
        cases,
        key=lambda c: (_SEVERITY_ORDER.get(c.severity, 99), c.due_date or dt.date.max, c.created_at),
    )
    return [get_case_out(db, company_id, case) for case in cases_sorted]

File: src/lib/api/cases.ts
import { calcDday } from '@/lib/dday';
import { useSessionStore } from '@/stores/sessionStore';
// 타입 전용 import(런타임에 지워짐) — SD-6이 CaseDetail의 새 필드를 mock CaseSheet와
// 같은 모양으로 맞춰(checkedItems/docs) 화면이 mock/real을 공유 로직으로 다루게 한다.
import type { CaseCheckedItem, CaseDoc, CaseDocStatus } from '@/mocks/fixtures';
import type { AgentStage, CaseCard, NextActionRef, WorkerRef } from '@/types';
import { apiFetch } from './client';

// GET /api/v1/cases 응답 DTO(백엔드 backend/app/schemas/case.py의 CaseOut 그대로, snake_case) —
// R2.3, NEXT_ROADMAP 2.3. briefings.ts가 이 타입·toCaseCard()를 그대로 재사용한다.
export interface WorkerRefDto {
  display_name: string;
  nationality: string;
  team: string | null;
}

export interface NextActionDto {
  action_id: string;
  label: string;
  state: string;
  requires_approval: boolean;
  kind: string;
}

export interface CaseDto {
  id: string;
  case_code: string;
  title: string;
  severity: string;
  state: string;
  agent_stage: string | null;
  due_date: string | null;
  approval_required: boolean;
  prepared_by: string;
  prepared_run_id: string | null;
  worker: WorkerRefDto | null;
  primary_action: NextActionDto | null;
  secondary_action: NextActionDto | null;
}

function toWorkerRef(dto: WorkerRefDto | null): WorkerRef | undefined {
  if (!dto) return undefined;
  return { displayName: dto.display_name, nationality: dto.nationality, team: dto.team ?? undefined, maskLevel: 'masked' };
}

// next_actions가 아직 없는 케이스(감지만 되고 액션 초안이 안 만들어진 경우)를 위한 안전한
// 기본값 — CaseCard.primaryAction/secondaryAction은 필수 필드라 null을 그대로 못 둔다.
// CSV/온보딩 유입 케이스의 기본 액션(lib/csvUpload.workerToCard)과 같은 패턴.
function toNextAction(
  dto: NextActionDto | null,
  fallback: { actionId: string; label: string; kind: NextActionRef['kind'] },
): NextActionRef {
  if (!dto) {
    return { actionId: fallback.actionId, label: fallback.label, state: 'ready', requiresApproval: false, kind: fallback.kind };
  }
  return {
    actionId: dto.action_id,
    label: dto.label,
    state: dto.state as NextActionRef['state'],
    requiresApproval: dto.requires_approval,
    kind: dto.kind as NextActionRef['kind'],
  };
}

// R2.3 배선 — 아직 GET /api/v1/cases 응답에 없는 CaseCard 필드(의도적 스코프 축소, 후속 확장
// 대상): stayExpiryDate(WorkerRefOut에 stay_expires_at 미포함) · missingDocCount · assignee ·
// evidenceCompleteness(next_actions/cases에 해당 개념 없음) · preparedRunRef(prepared_run_id는
// 있으나 evidenceRef "#4788" 표기로 변환할 근거[runs.anchor_event_no]가 아직 안 내려온다).
// 여기서 값을 지어내지 않고 undefined로 남긴다 — 화면은 이미 이 필드들을 옵셔널로 다룬다.
export function toCaseCard(dto: CaseDto): CaseCard {
  return {
    caseId: dto.id,
    caseCode: dto.case_code,
    title: dto.title,
    workerRef: toWorkerRef(dto.worker),
    severity: dto.severity as CaseCard['severity'],
    // 실 API 모드는 데모 고정일(DEMO_TODAY)이 아니라 진짜 오늘을 기준으로 dDay를 계산한다 —
    // 실서버 데이터는 데모 세계관 날짜에 묶이지 않는다(lib/dday.ts의 "기준일 주입" 원칙은
    // 테스트 결정성을 위한 것이지, 여기서 new Date()를 쓰는 것과 상충하지 않는다).
    dDay: dto.due_date ? calcDday(dto.due_date, new Date()) : undefined,
    state: dto.state as CaseCard['state'],
    agentStage: (dto.agent_stage as AgentStage | null) ?? undefined,
    approvalRequired: dto.approval_required,
    primaryAction: toNextAction(dto.primary_action, { actionId: `${dto.id}-detail`, label: '상세 보기', kind: 'detail' }),
    secondaryAction: toNextAction(dto.secondary_action, { actionId: `${dto.id}-confirm`, label: '케이스 확인 완료', kind: 'confirm' }),
    preparedBy: dto.prepared_by as CaseCard['preparedBy'],
  };
}

export async function fetchCases(): Promise<CaseCard[]> {
  const token = useSessionStore.getState().token ?? undefined;
  const dtos = await apiFetch<CaseDto[]>('/api/v1/cases', { token });
  return dtos.map(toCaseCard);
}

// GET /api/v1/cases/{case_id} 응답 DTO(R2.4) — CaseOut 필드 전부 + ApprovePage 전용 필드.
export interface PendingApprovalChecklistItemDto {
  key: string;
  label: string;
  checked: boolean;
}

export interface PendingApprovalDto {
  id: string;
  action_id: string;
  checklist: PendingApprovalChecklistItemDto[] | null;
  requested_at: string;
}

// SD-6 — backend/app/schemas/case.py CheckedItemOut/WorkerDocumentOut 그대로.
export interface CheckedItemDto {
  label: string;
  value: string;
}

export interface WorkerDocumentDto {
  doc_type: string;
  status: string;
  due_date: string | null;
  expires_at: string | null;
}

export interface CaseDetailDto extends CaseDto {
  usable_citation_count: number;
  guard_note: string | null;
  pending_approval: PendingApprovalDto | null;
  checked_items: CheckedItemDto[];
  next_wake: string | null;
  documents: WorkerDocumentDto[];
}

export interface PendingApprovalChecklistItem {
  key: string;
  label: string;
  checked: boolean;
}

export interface PendingApproval {
  id: string;
  actionId: string;
  checklist: PendingApprovalChecklistItem[] | null;
  requestedAt: string;
}

// ApprovePage/CaseReviewPage/CaseWorkbench 등이 real 모드에서 CASE_SHEETS(mock) 대신 쓰는
// 필드만 뽑는다 — 카드 본체는 caseStore(fetchCases가 이미 채움)를 그대로 쓴다(중복 조립 방지).
// checkedItems/docs/nextWake는 SD-6(plans/SEED_DESIGN_2026-07-20.md Part B5(c)) — 필드명은
// mock CaseSheet(mocks/fixtures.ts)과 일부러 맞췄다(화면이 동일 렌더 로직을 재사용할 수 있게).
export interface CaseDetail {
  usableCitationCount: number;
  guardNote: string | null;
  pendingApproval: PendingApproval | null;
  checkedItems: CaseCheckedItem[];
  nextWake?: string;
  docs: CaseDoc[];
}

function toPendingApproval(dto: PendingApprovalDto | null): PendingApproval | null {
  if (!dto) return null;
  return {
    id: dto.id,
    actionId: dto.action_id,
    checklist: dto.checklist?.map((item) => ({ key: item.key, label: item.label, checked: item.checked })) ?? null,
    requestedAt: dto.requested_at,
  };
}

// mock CaseDoc.statusLabel은 디자인 원문이 케이스마다 다르게 써 둔 자유 문구("05/12 확인" 등,
// GOTCHAS §3 배지-라벨 병기 관례)라 서버가 그대로 재현할 근거가 없다 — real 모드는 상태값
// 자체를 사람이 읽는 짧은 라벨로만 표시한다(정직한 단순화, CaseDocStatus 6종 전부 커버).
const DOC_STATUS_LABEL: Record<CaseDocStatus, string> = {
  missing: '누락',
  requested: '요청됨',
  received: '확보',
  expiring: '만료 임박',
  company_check: '사내 확인 중',
  pending: '대기',
};

function toCaseDoc(dto: WorkerDocumentDto): CaseDoc {
  const status = dto.status as CaseDocStatus;
  return { name: dto.doc_type, status, statusLabel: DOC_STATUS_LABEL[status] ?? dto.status };
}

export async function fetchCaseDetail(caseId: string): Promise<CaseDetail> {
  const token = useSessionStore.getState().token ?? undefined;
  const dto = await apiFetch<CaseDetailDto>(`/api/v1/cases/${caseId}`, { token });
  return {
    usableCitationCount: dto.usable_citation_count,
    guardNote: dto.guard_note,
    pendingApproval: toPendingApproval(dto.pending_approval),
    checkedItems: dto.checked_items.map((item) => ({ label: item.label, value: item.value })),
    nextWake: dto.next_wake ?? undefined,
    docs: dto.documents.map(toCaseDoc),
  };
}
